File: src/api/models/tabs.py
"""Tab 级聚合响应模型（前端每 tab 一个 API）

切片原则：
- 每个 tab 端点只返回该 tab 展示所需的数据；
- 指标类数据一律为服务端聚合后的复合数据，底层原始行（如实体候选
  span 明细）不出现在 tab 响应中，展示主体（LLM 报告/图快照/伏笔线程）
  按展示对象原样透传；
- 沿用仓库事实标准：run_id + 溯源块 + unavailable_reason 三件套，
  不做统一 envelope。
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from src.api.models.graph import GraphMetricsResponse, GraphSnapshotResponse, KeywordItem
from src.api.models.responses import (
    ChapterMetricsResponse,
    ChapterTopicDistribution,
    CharacterStats,
    CharacterStatsAggregate,
    DiagnosisResult,
    EmotionStats,
    EmotionTrendWindow,
    NarrativeStructureStats,
    ParagraphCurvePoint,
    StyleStats,
    TopicDistributionEntry,
    TopicInfo,
    TopicModelMetaInfo,
)


class TopicsOverviewTabResponse(BaseModel):
    """主题总览 tab：主题词 + 全书/章节完整分布 + TextRank 关键词 + 诊断主题标签

    - topics/distribution/chapters 缺模型契约行时为空且 unavailable_reason 非空
    - keywords 口径独立于 LDA（TextRank），不可用原因单独回显
    - topic_labels 为诊断切片（LLM 主题命名，按 topic_id 顺序对齐 topics）
    """

    run_id: str
    model: TopicModelMetaInfo | None = None
    topics: list[TopicInfo] = Field(default_factory=list)
    distribution: list[TopicDistributionEntry] | None = None
    chapters: list[ChapterTopicDistribution] = Field(default_factory=list)
    keywords: list[KeywordItem] = Field(default_factory=list)
    topic_labels: list[str] | None = None
    unavailable_reason: str | None = None
    keyword_unavailable_reason: str | None = None


class EntitySurfaceCount(BaseModel):
    """高频实体名聚合（surface_text + 归一类型 group-by 计数）"""

    surface_text: str
    entity_type: str
    count: int


class LinguisticEntitiesTabResponse(BaseModel):
    """实体与短语 tab：实体类型计数 + 高频实体名 + 固定短语统计

    不含实体候选 span 明细（paragraph_id/字符区间），仅回聚合统计；
    unavailable_reason 表示语言阶段不可用；已运行但没有实体候选时仍展示短语统计。
    """

    run_id: str
    count_by_type: dict[str, int] = Field(default_factory=dict)
    surface_top: list[EntitySurfaceCount] = Field(default_factory=list)
    total_char_count: int = 0
    metric_hit_count: int = 0
    fixed_phrase_density: float | None = None
    four_char_candidate_count: int = 0
    total_hits: int = 0
    unavailable_reason: str | None = None


class DashboardTabResponse(BaseModel):
    """仪表盘 tab：原有 8 个并发请求合并为一次拉取

    各 section 复用既有端点的响应模型；情绪趋势固定全书窗口 20 段。
    """

    run_id: str
    narrative_structure: NarrativeStructureStats | None = None
    emotion_stats: EmotionStats | None = None
    character_stats: CharacterStatsAggregate | None = None
    style_stats: StyleStats | None = None
    chapter_metrics: ChapterMetricsResponse | None = None
    topics: list[TopicInfo] = Field(default_factory=list)
    diagnosis: DiagnosisResult | None = None
    emotion_trend: list[EmotionTrendWindow] = Field(default_factory=list)


class RhythmTabResponse(BaseModel):
    """节奏张力 tab：段落曲线（max_points 展示降采样）+ 叙事结构高潮参数"""

    run_id: str
    curves: list[ParagraphCurvePoint] = Field(default_factory=list)
    narrative_structure: NarrativeStructureStats | None = None


class CharacterFunctionTabResponse(BaseModel):
    """功能与焦点 tab：角色功能分布 + 诊断焦点结构切片

    characters 沿用 /characters 的别名归并与焦点对齐口径。
    """

    run_id: str
    characters: list[CharacterStats] = Field(default_factory=list)
    focus_structure: Literal["single", "dual", "ensemble"] | None = None
    focus_characters: list[str] | None = None
    arc_scores: dict[str, float] | None = None


class CharacterAppearance(BaseModel):
    """图谱页登场次数切片（仅 name + appearance_count）"""

    name: str
    appearance_count: int


class GraphNetworkTabResponse(BaseModel):
    """关系图谱/快照概览 tab：图快照 + 登场次数 + 图算法指标 + 变化总数

    - snapshot/graph_metrics 各自缺失时为 None 并回显 unavailable_reason，
      不以空快照伪造；
    - change_total 为图谱变化总条数（供快照概览计数，不回变化明细）。
    """

    run_id: str
    snapshot: GraphSnapshotResponse | None = None
    character_appearances: list[CharacterAppearance] = Field(default_factory=list)
    graph_metrics: GraphMetricsResponse | None = None
    change_total: int = 0
    unavailable_reason: str | None = None

File: frontend/src/mocks/handlers/tabs.ts
/**
 * MSW Handler — Tab 级聚合端点（/tabs/*）
 *
 * 组合既有 mock 数据工厂，保持与真实后端 tabs 响应相同的切片口径：
 * 指标类数据为聚合值，展示主体（图快照/诊断/伏笔线程）原样透传。
 */
import { http, HttpResponse, delay } from "msw";
import {
  createCharacters,
  createChapterMetrics,
  createDiagnosis,
  createEmotionStats,
  createEmotionTrendWindows,
  createGraph,
  createGraphChangesPage,
  createNarrativeStructure,
  createParagraphCurves,
  createCharacterStats,
  createStyleStats,
  createTopics,
  taskDb,
} from "../data";

const BASE = import.meta.env.VITE_API_BASE_URL || "";
const READABLE_TASK_STATUSES = new Set(["completed"]);

async function checkTaskReady(novelId: string, taskId: string): Promise<Response | null> {
  const tasks = taskDb.get(novelId) ?? [];
  const task = tasks.find((t) => t.task_id === taskId);

  if (!task) {
    return HttpResponse.json({ detail: "任务不存在" }, { status: 404 });
  }
  if (!READABLE_TASK_STATUSES.has(task.status)) {
    return HttpResponse.json(
      {
        detail: `分析未完成，当前状态: ${task.status}`,
        error_type: "AnalysisNotCompleteError",
        status_code: 400,
        run_status: task.status,
      },
      { status: 400 }
    );
  }
  return null;
}

function buildTopicModelMeta(numTopics: number) {
  return {
    model_key: "gensim-lda",
    library_version: "4.4.0",
    pipeline_version: "1.0",
    num_topics: numTopics,
    artifact_key: "models/topic/mock-run",
  };
}

// GET /api/novels/:novelId/tabs/dashboard
export const dashboardTabHandler = http.get(
  `${BASE}/api/novels/:novelId/tabs/dashboard`,
  async ({ request, params }) => {
    const { novelId } = params;
    const taskId = new URL(request.url).searchParams.get("task_id") ?? "";
    const err = await checkTaskReady(novelId as string, taskId);
    if (err) return err;

    await delay(600);
    return HttpResponse.json({
      run_id: taskId,
      narrative_structure: createNarrativeStructure(),
      emotion_stats: createEmotionStats(),
      character_stats: createCharacterStats(),
      style_stats: createStyleStats(),
      chapter_metrics: createChapterMetrics(),
      topics: createTopics(),
      diagnosis: createDiagnosis(),
      emotion_trend: createEmotionTrendWindows(20, null),
    });
  }
);

// GET /api/novels/:novelId/tabs/rhythm
export const rhythmTabHandler = http.get(
  `${BASE}/api/novels/:novelId/tabs/rhythm`,
  async ({ request, params }) => {
    const { novelId } = params;
    const url = new URL(request.url);
    const taskId = url.searchParams.get("task_id") ?? "";
    const err = await checkTaskReady(novelId as string, taskId);
    if (err) return err;

    await delay(400);
    const maxPoints = Number(url.searchParams.get("max_points"));
    const count = Number.isFinite(maxPoints) && maxPoints > 0 ? Math.min(maxPoints, 5000) : 800;
    return HttpResponse.json({
      run_id: taskId,
      curves: createParagraphCurves(count),
      narrative_structure: createNarrativeStructure(),
    });
  }
);

// GET /api/novels/:novelId/tabs/character-function
export const characterFunctionTabHandler = http.get(
  `${BASE}/api/novels/:novelId/tabs/character-function`,
  async ({ request, params }) => {
    const { novelId } = params;
    const taskId = new URL(request.url).searchParams.get("task_id") ?? "";
    const err = await checkTaskReady(novelId as string, taskId);
    if (err) return err;

    await delay(300);
    const diagnosis = createDiagnosis();
    return HttpResponse.json({
      run_id: taskId,
      characters: createCharacters(),
      focus_structure: diagnosis.focus_structure ?? null,
      focus_characters: diagnosis.focus_characters ?? null,
      arc_scores: diagnosis.arc_scores ?? null,
    });
  }
);

// GET /api/novels/:novelId/tabs/graph-network
export const graphNetworkTabHandler = http.get(
  `${BASE}/api/novels/:novelId/tabs/graph-network`,
  async ({ request, params }) => {
    const { novelId } = params;
    const url = new URL(request.url);
    const taskId = url.searchParams.get("task_id") ?? "";
    const err = await checkTaskReady(novelId as string, taskId);
    if (err) return err;

    await delay(400);
    const characters = createCharacters();
    const changeTotal = createGraphChangesPage(null, 1).page_info.total;
    return HttpResponse.json({
      run_id: taskId,
      snapshot: createGraph(),
      character_appearances: characters.map((character) => ({
        name: character.name,
        appearance_count: character.appearance_count,
      })),
      graph_metrics: {
        run_id: taskId,
        unavailable_reason: null,
        algorithm: { version: "graph-metrics-v1" },
        pagerank: Object.fromEntries(
          characters.slice(0, 8).map((character, index) => [character.name, Number((0.3 - index * 0.03).toFixed(6))])
        ),
        hits: { authority: {}, hub: {} },
        communities: { community_ids: {}, modularity: 0.42, interpretation: "structural_community_only" },
      },
      change_total: changeTotal,
      unavailable_reason: null,
    });
  }
);

// GET /api/novels/:novelId/tabs/topics-overview
export const topicsOverviewTabHandler = http.get(
  `${BASE}/api/novels/:novelId/tabs/topics-overview`,
  async ({ request, params }) => {
    const { novelId } = params;
    const taskId = new URL(request.url).searchParams.get("task_id") ?? "";
    const err = await checkTaskReady(novelId as string, taskId);
    if (err) return err;

    await delay(400);
    const topics = createTopics();
    return HttpResponse.json({
      run_id: taskId,
      model: buildTopicModelMeta(topics.length),
      topics,
      distribution: topics.map((topic) => ({ topic_id: topic.topic_id, weight: topic.weight })),
      chapters: [1, 2, 3].map((chapterId) => ({
        chapter_id: chapterId,
        chapter_sequence: chapterId,
        chapter_title: `第${chapterId}章`,
        token_total: 1200,
        distribution: topics.map((topic, index) => ({
          topic_id: topic.topic_id,
          weight: Number(Math.max(0, topic.weight + (index % 2 === 0 ? 0.02 : -0.02) * chapterId).toFixed(6)),
        })),
      })),
      keywords: [
        { word: "宗门", score: 0.0812 },
        { word: "灵脉", score: 0.0643 },
        { word: "剑意", score: 0.0521 },
        { word: "长老", score: 0.0417 },
        { word: "秘境", score: 0.0388 },
      ],
      unavailable_reason: null,
      keyword_unavailable_reason: null,
    });
  }
);

// GET /api/novels/:novelId/tabs/linguistic-entities
export const linguisticEntitiesTabHandler = http.get(
  `${BASE}/api/novels/:novelId/tabs/linguistic-entities`,
  async ({ request, params }) => {
    const { novelId } = params;
    const taskId = new URL(request.url).searchParams.get("task_id") ?? "";
    const err = await checkTaskReady(novelId as string, taskId);
    if (err) return err;

    await delay(300);
    return HttpResponse.json({
      run_id: taskId,
      count_by_type: { person: 186, location: 74, organization: 31, item: 22 },
      surface_top: [
        { surface_text: "林渡", entity_type: "person", count: 92 },
        { surface_text: "顾霜", entity_type: "person", count: 61 },
        { surface_text: "萧遥", entity_type: "person", count: 33 },
        { surface_text: "青云宗", entity_type: "organization", count: 18 },
        { surface_text: "落霞谷", entity_type: "location", count: 11 },
      ],
      total_char_count: 186000,
      metric_hit_count: 47,
      fixed_phrase_density: 0.2527,
      four_char_candidate_count: 12,
      total_hits: 59,
      unavailable_reason: null,
    });
  }
);
